streams/src/test/java/org/apache/kafka/streams/processor/TimestampExtractorTest.java

```diff
@@ -0,0 +1,136 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.record.TimestampType;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class TimestampExtractorTest {
+
+    @Test
+    public void extractMetadataTimestamp() {
+        final long metadataTimestamp = 42;
+
+        final TimestampExtractor[] extractors = new TimestampExtractor[]{
+            new FailOnInvalidTimestamp(),
```

Comment:
`FailOnInvalidTimestamp`, `LogAndSkipOnInvalidTimestamp`, and `UsePreviousTimeOnInvalidTimestamp` extract record metadata timestamp and are thus similar -- I think it's easier to do a single test class to avoid code duplication (c.f. `#extractMetadataTimestamp()`).